tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs

```diff
@@ -0,0 +1,481 @@
+                        {
+                            issueLogger.LogBreakingChangeIssue(
+                                cmdlet: cmdlet,
+                                severity: 0,
+                                problemId: ProblemIds.BreakingChangeProblemId.ChangedElementType,
+                                description: string.Format(Properties.Resources.ChangedElementTypeDescription,
+                                    oldProperty, oldTypeMetadata.ElementType, newTypeMetadata.ElementType),
+                                remediation: string.Format(Properties.Resources.ChangedElementTypeRemediation,
+                                    oldProperty, oldTypeMetadata.ElementType));
+                        }
+
+                        continue;
+                    }
+
+                    // Check if the type is a generic
+                    if (oldTypeMetadata.GenericTypeArguments.Count > 0 && newTypeMetadata.GenericTypeArguments.Count > 0)
+                    {
+                        // Check if the generic type has changed
+                        if (oldTypeMetadata.Name.Equals(newTypeMetadata.Name, StringComparison.OrdinalIgnoreCase))
+                        {
+                            // Check if the number of generic type arguments is the same
+                            if (oldTypeMetadata.GenericTypeArguments.Count == newTypeMetadata.GenericTypeArguments.Count)
+                            {
+                                // For each element in the generic type arguments list, make sure that the types
+                                // are the same
+                                for (int idx = 0; idx < oldTypeMetadata.GenericTypeArguments.Count; idx++)
+                                {
+                                    if (oldTypeMetadata.GenericTypeArguments[idx].Equals(newTypeMetadata.GenericTypeArguments[idx], StringComparison.OrdinalIgnoreCase))
+                                    {
+                                        // If we have not previously seen this generic type argument,
+                                        // run this method on the type
+                                        if (!_typeSet.Contains(oldTypeMetadata.GenericTypeArguments[idx]))
+                                        {
+                                            _typeSet.Add(oldTypeMetadata.GenericTypeArguments[idx]);
+
+                                            var oldElementType = _oldTypeDictionary[oldTypeMetadata.GenericTypeArguments[idx]];
+                                            var newElementType = _newTypeDictionary[newTypeMetadata.GenericTypeArguments[idx]];
+
+                                            CompareTypeMetadata(cmdlet, oldElementType, newElementType, issueLogger);
+                                        }
+                                    }
+                                    // If the generic type arguments aren't the same, log an issue
+                                    else
```

Comment:
so, if they aren't the same, don't we want to compare the properties recirsively?